### PR TITLE
Add segmentation and bounding box sensor types

### PR DESCRIPTION
**Pull request by AmrElsersy**
# 🎉 New feature

related to [#134](https://github.com/ignitionrobotics/ign-sensors/issues/134) [#135](https://github.com/ignitionrobotics/ign-sensors/issues/135)

## Summary
Added new types for SDF::Sensor to support Segmentation & BoundingBox Sensors

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] Code check passed (In source directory, run `sh tools/code_check.sh`)
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review

### CODE REVIEW

**Review by ahcorde — APPROVED** [review]

**Review by adlarkin — CHANGES_REQUESTED** [review]
Looks good so far, but we should do a few more things before merging this:
1. Can you update `test/integration/link_dom.cc` to test the code path from XML to C++ for the new sensor types? Here's an example you can follow: https://github.com/osrf/sdformat/blob/sdformat11_11.1.0/test/integration/link_dom.cc#L337-L346
2. Can you update the sensors unit test? You'll need to add the new types here: https://github.com/osrf/sdformat/blob/sdformat11_11.1.0/src/Sensor_TEST.cc#L243-L285
3. Can you update the sensor documentation to include these new types? You'll want to update `sdf/1.9/sensor.sdf`: https://github.com/osrf/sdformat/blob/1a779837277034ec45f67cc9f1ea9a47bf4bfd9a/sdf/1.9/sensor.sdf#L11-L32
4. Can you change the name of the PR to say something like "_add segmentation and bounding box sensor types_"? The current PR name is generic, and doesn't really give any information about the changes being made.

**Comment by chapulina** [comment]
I noticed that the bounding box camera implementation in https://github.com/ignitionrobotics/ign-rendering/pull/334 accepts some configuration: visible / full. Do we want users to be able to configure that through SDF? Are there any other configurations we may want to add here for either sensor?

**Comment by adlarkin** [comment]
> I noticed that the bounding box camera implementation in [ignitionrobotics/ign-rendering#334](https://github.com/ignitionrobotics/ign-rendering/pull/334) accepts some configuration: visible / full. Do we want users to be able to configure that through SDF? Are there any other configurations we may want to add here for either sensor?

I think it's likely that we will want users to configure `visible/full` through SDF, and there's a chance that there are other configurations we'd want to add to these sensors as well. I think that the best approach is to finish reviewing/iterating over all of the other bounding box/segmentation PRs before merging this PR - that way, if we find that the SDF configuration needs to be modified at all, we can do so before merging.

**Comment by AmrElsersy** [comment]
@chapulina @adlarkin 
I do these configurations in ign-sensors, by getting them through the SDF
like that: https://github.com/ignitionrobotics/ign-sensors/pull/133/files#diff-3a47a41297adece53d9e25cb71940e30d655d1a491020c1a8fe918a4f1a95ac2R146

**Comment by adlarkin** [comment]
> I do these configurations in ign-sensors, by getting them through the SDF
> like that: https://github.com/ignitionrobotics/ign-sensors/pull/133/files#diff-3a47a41297adece53d9e25cb71940e30d655d1a491020c1a8fe918a4f1a95ac2R146
> 
> ```c++
> std::string type = sdfElement->Get<std::string>("segmentation_type");
> ```
> 
> is that a bad way to do that ? :D

I think what you're doing in `ign-sensors` is correct, but we want to make sure that we document these configurations in SDF so that users know there are additional/unique parameters that they can configure for the segmentation and bounding box sensors. So, we may need to add files like `segmentation_camera.sdf` and `boundingbox_camera.sdf` (I'm not sure what the naming conventions should be for these files - do you know, @chapulina?) to the `sdf/1.9/` directory, and then include these files in `sdf/1.9/sensors.sdf`. These new `*.sdf` files would specify the elements that are unique to the segmentation and bounding box sensors. Here's an example with the air pressure sensor:

* [`air_pressure.sdf`](https://github.com/osrf/sdformat/blob/1a779837277034ec45f67cc9f1ea9a47bf4bfd9a/sdf/1.9/air_pressure.sdf) - this file is what specifies the elements that are unique to an air pressure sensor.
* Including `air_pressure.sdf` in `sensors.sdf`: https://github.com/osrf/sdformat/blob/1a779837277034ec45f67cc9f1ea9a47bf4bfd9a/sdf/1.9/sensor.sdf#L55

**Comment by chapulina** [comment]
> I'm not sure what the naming conventions should be for these files - do you know, @chapulina?

Just looking at the existing files, your suggestions look good

**Comment by AmrElsersy** [comment]
I will add the sdf examples in the ign-sensors repo, as we discussed ashton, so I think this PR should be ready to review.

File: test/sdf/sensors.sdf [hunk 1]
```diff
@@ -308,7 +308,7 @@
         </lidar>
       </sensor>
 
- 
+
```
Comment (adlarkin) [inline test/sdf/sensors.sdf:311]: nit: remove this extra empty line

File: test/sdf/sensors.sdf [hunk 2]
```diff
@@ -435,6 +435,121 @@
...
+              <cx>162</cx>
+              <cy>124</cy>
+              <s>1.2</s>
+            </intrinsics>
+          </lens>
+        </camera>
+
+      </sensor>
+
+
```
Comment (adlarkin) [inline test/sdf/sensors.sdf:552]:
```suggestion

```
nit: only have one new line separating the sensors instead of two new lines

**Review by adlarkin — COMMENTED** [review]
Can you also address the changes mentioned in https://github.com/ignitionrobotics/sdformat/pull/592#issuecomment-863380707?

File: test/integration/link_dom.cc [hunk 3]
```diff
@@ -345,6 +345,30 @@ TEST(DOMLink, Sensors)
...
+  EXPECT_EQ(sdf::SensorType::SEGMENTATION_CAMERA, segmentationSensor->Type());
+  EXPECT_EQ(ignition::math::Pose3d(37, 38, 39, 0, 0, 0),
+            segmentationSensor->RawPose());
+  const sdf::Camera *segmentationCameraSensor =
+    segmentationSensor->CameraSensor();
+  ASSERT_NE(nullptr, segmentationCameraSensor);
+  EXPECT_EQ("my_segmentation_camera", segmentationCameraSensor->Name());
+
+  // Get the boundingbox sensor
+  const sdf::Sensor *boundingboxSensor = link->SensorByName("boundingbox_sensor");
```
Comment (adlarkin) [inline test/integration/link_dom.cc:362]: Codecheck is failing in this line:
```
./test/integration/link_dom.cc:362:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
```

**Comment by adlarkin** [comment]
Tests in CI are failing, @AmrElsersy can you address the failures?

File: sdf/1.9/segmentation_camera.sdf [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    <description>
+      The segmentation type of the segmentation camera, the valid options:
+        - semantic: Semantic segmentation which provides a gray image which its pixels
+                    have the label value of the objects, and a colored image which its pixels
+                    have a unique color for each label.
+
+        - panoptic | instance: Panoptic segmentation which provides an image where each pixel
+                               has 1 channel for label value of the object and 2 channels for the
+                               number of the instances of that label, and a colored image which its
+                               pixels have a unique color for each instance.
```
Comment (adlarkin) [inline sdf/1.9/segmentation_camera.sdf:14]:
```suggestion
        - panoptic | instance: Panoptic segmentation, which provides 2 images:
                               1. A 3-channel image, where each pixel has 1 channel for the label value of the object and 2 channels for the number of the instances of that label
                               2. A colored image, with the pixel values being a unique color for each instance
```

@chapulina or @azeey: for the 3-channel image described here, should we specify which channel is associated with the label and which 2 channels are associated with the instance count? Or should we not define this here so that it can be defined by downstream users of SDF?

File: test/integration/link_dom.cc [hunk 5]
```diff
@@ -345,6 +345,31 @@ TEST(DOMLink, Sensors)
...
+    link->SensorByName("segmentation_sensor");
+  ASSERT_NE(nullptr, segmentationSensor);
+  EXPECT_EQ("segmentation_sensor", segmentationSensor->Name());
+  EXPECT_EQ(sdf::SensorType::SEGMENTATION_CAMERA, segmentationSensor->Type());
+  EXPECT_EQ(ignition::math::Pose3d(37, 38, 39, 0, 0, 0),
+            segmentationSensor->RawPose());
+  const sdf::Camera *segmentationCameraSensor =
+    segmentationSensor->CameraSensor();
+  ASSERT_NE(nullptr, segmentationCameraSensor);
+  EXPECT_EQ("my_segmentation_camera", segmentationCameraSensor->Name());
```
Comment (adlarkin) [inline test/integration/link_dom.cc:359]: Can we also test the unique elements for a segmentation sensor (`segmentation_type`)? You'll need to add the unique elements to `test/sdf/sensors.sdf`

File: test/integration/link_dom.cc [hunk 6]
```diff
@@ -345,6 +345,31 @@ TEST(DOMLink, Sensors)
...
+  const sdf::Sensor *boundingboxSensor =
+    link->SensorByName("boundingbox_sensor");
+  ASSERT_NE(nullptr, boundingboxSensor);
+  EXPECT_EQ("boundingbox_sensor", boundingboxSensor->Name());
+  EXPECT_EQ(sdf::SensorType::BOUNDINGBOX_CAMERA, boundingboxSensor->Type());
+  EXPECT_EQ(ignition::math::Pose3d(37, 38, 39, 0, 0, 0),
+            boundingboxSensor->RawPose());
+  const sdf::Camera *boundingboxCamSensor = boundingboxSensor->CameraSensor();
+  ASSERT_NE(nullptr, boundingboxCamSensor);
+  EXPECT_EQ("my_boundingbox_camera", boundingboxCamSensor->Name());
```
Comment (adlarkin) [inline test/integration/link_dom.cc:371]: Can we also test the unique elements for a boundingbox sensor (`box_type`)? You'll need to add the unique elements to `test/sdf/sensors.sdf`

File: sdf/1.9/bounding_box_camera.sdf [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  <element name="box_type" type="string" default="2d" required="0">
+    <description>
+      The boundingbox type of the boundingbox camera, the valid options:
+        - 2d | visible_2d | visible_box_2d: a visible 2d box mode which provides and axis aligned 2d boxes
+                                            on the visible parts of the objects
+
+        - full_2d | full_box_2d: a full 2d box mode which provides an axis aligned 2d boxes that fills the
+                                 object dimentions even if it had an occluded part
+
+        - 3d: a 3d mode which provides oriented 3d boxes
```
[the remark this answers is not in the document]
Comment (AmrElsersy) [reply sdf/1.9/bounding_box_camera.sdf:13]: Full, because they are a real 3d position and dimentions ... there is no meaning in visible concept here ... no body will suppose the visible concept in the 3d

File: test/integration/link_dom.cc (hunk 5 above)
Comment (AmrElsersy) [reply test/integration/link_dom.cc:359]: We discussed that before in the above comments, we agreed that the approach of accessing that types from ign-sensors (via the SDF API) should be fine

here
> I do these configurations in ign-sensors, by getting them through the SDF
> like that: https://github.com/ignitionrobotics/ign-sensors/pull/133/files#diff-3a47a41297adece53d9e25cb71940e30d655d1a491020c1a8fe918a4f1a95ac2R146
> 
> ```c++
> std::string type = sdfElement->Get<std::string>("segmentation_type");
> ```
> 
> is that a bad way to do that ? :D

to add this unique elements, i think I should create a new SDF sensor type for the semgnetation type for example

File: test/integration/link_dom.cc (hunk 6 above)
Comment (AmrElsersy) [reply test/integration/link_dom.cc:371]: same

File: test/integration/link_dom.cc (hunk 5 above)
Comment (adlarkin) [reply test/integration/link_dom.cc:372]: Hmm okay so after looking through the SDF repo for a bit, I actually think that there's a better way than `sdfElement->Get<std::string>` for retrieving the unique element that's associated with a sensor. I think that we should follow the `depth_camera` as an example for creating a camera with a few custom parameters.

If we look at the `depth_camera`, it has a few unique elements on top of all the other `camera` elements: https://github.com/ignitionrobotics/sdformat/blob/sdformat11_11.2.2/sdf/1.8/camera.sdf#L47-L63

These unique elements are accessed via getters/setters in the `sdf::Camera` class:
* https://github.com/ignitionrobotics/sdformat/blob/sdformat11_11.2.2/include/sdf/Camera.hh#L139-L153
* https://github.com/ignitionrobotics/sdformat/blob/sdformat11_11.2.2/include/sdf/Camera.hh#L163-L189

Here's an example of how this looks in sdf (notice `depth_camera` being used in lines 54-59):
* https://github.com/osrf/subt/blob/bff4f4898fa9c2aabeef1b69833d7cf04ee356fa/submitted_models/ctu_cras_norlab_absolem_sensor_config_1/urdf/realsense.gazebo.xacro#L25-L69

---

So, I think we should actually do the following:
1. Add getters/setters to the `sdf::Camera` class for the segmentation and bounding box types
2. Update the camera SDF docs (`sdf/1.9/camera.sdf`) to include unique elements for segmentation and bounding box cameras (see the [depth camera elements](https://github.com/ignitionrobotics/sdformat/blob/sdformat11_11.2.2/sdf/1.8/camera.sdf#L47-L63) as an example). This shouldn't be too difficult - just move the content of your new `sdf/1.9/*.sdf` files over to `sdf/1.9/camera.sdf`. Once everything is moved to `sdf/1.9/camera.sdf`, you can delete the other `sdf/1.9/*.sdf` files that you created and also remove lines [60](https://github.com/ignitionrobotics/sdformat/blob/178ae03abe3f93d1a9eb89b49e68093c52b20b3d/sdf/1.9/sensor.sdf#L60) and [72](https://github.com/ignitionrobotics/sdformat/blob/178ae03abe3f93d1a9eb89b49e68093c52b20b3d/sdf/1.9/sensor.sdf#L72) from `sdf/1.9/sensors.sdf` 
3. Update SDF files that use these sensors (both in this PR and other PRs) to call the unique elements (`segmentation_type` and `box_type`) inside of the `<camera>` element instead of the `<sensor>` element, since we are adding this functionality to the `sdf::Camera` class
4. Update the `sdf::Camera` unit test to cover these new getters/setters (example from the `depth_camera` elements: https://github.com/ignitionrobotics/sdformat/blob/sdformat11_11.2.2/src/Camera_TEST.cc#L45-L55)
5. Double-check all other related PRs that use these sensors and update them to use the newly added `sdf::Camera` APIs as needed

Here's the commit that adds a few unique elements to `depth_camera` that you can use as a reference/example: https://github.com/ignitionrobotics/sdformat/commit/1cbdfe4121f9b4bb51ddfeb0bac9bd6a0153e2a6#

After you make these changes, we should no longer need to use `sdfElement->Get<std::string>` for getting the segmentation and bounding box type elements (it will now be a part of the `sdf::Camera` API). Let me know if you have any questions!

File: test/integration/link_dom.cc (hunk 6 above)
Comment (adlarkin) [reply test/integration/link_dom.cc:384]: See https://github.com/ignitionrobotics/sdformat/pull/592#discussion_r688228488

File: sdf/1.9/camera.sdf [hunk 8]
```diff
@@ -62,6 +62,33 @@
     </element>
   </element> <!-- End depth_camera -->
 
+  <element name="segmentation_type" type="string" default="semantic" required="0">
+    <description>
+      The segmentation type of the segmentation camera. Valid options are:
+        - semantic: Semantic segmentation, which provides 2 images:
+                    1. A gray image, with the pixel values representing the label of an object
```
Comment (adlarkin) [inline sdf/1.9/camera.sdf:69]:
```suggestion
                    1. A grayscale image, with the pixel values representing the label of an object
```

File: sdf/1.9/camera.sdf [hunk 9]
```diff
@@ -62,6 +62,33 @@
...
+                    2. A colored image, with the pixel values being a unique color for each label
+
+        - panoptic | instance: Panoptic segmentation which provides an image where each pixel
```
Comment (adlarkin) [inline sdf/1.9/camera.sdf:72]:
```suggestion
        - panoptic | instance: Panoptic segmentation, which provides an image where each pixel
```

File: src/Camera_TEST.cc [hunk 10]
```diff
@@ -62,6 +62,18 @@ TEST(DOMCamera, Construction)
...
+  EXPECT_TRUE(cam.HasSegmentationType());
+
+  cam.SetBoundingBoxType("3d");
+  EXPECT_EQ("3d", cam.BoundingBoxType());
+  EXPECT_TRUE(cam.HasBoundingBoxType());
+
+  cam.SetDepthNearClip(0.2);
+  EXPECT_DOUBLE_EQ(0.2, cam.DepthNearClip());
+  EXPECT_TRUE(cam.HasDepthNearClip());
+
```
Comment (adlarkin) [inline src/Camera_TEST.cc:76]: This doesn't seem related to this PR, can it be removed?

File: src/Camera_TEST.cc [hunk 11]
```diff
@@ -62,6 +62,18 @@ TEST(DOMCamera, Construction)
   cam.SetFarClip(200.2);
   EXPECT_DOUBLE_EQ(200.2, cam.FarClip());
 
+  cam.SetSegmentationType("semantic");
+  EXPECT_EQ("semantic", cam.SegmentationType());
+  EXPECT_TRUE(cam.HasSegmentationType());
+
+  cam.SetBoundingBoxType("3d");
+  EXPECT_EQ("3d", cam.BoundingBoxType());
+  EXPECT_TRUE(cam.HasBoundingBoxType());
```
Comment (adlarkin) [inline src/Camera_TEST.cc:71]: Can you also add test coverage for the `SetHasSegmentationType` and `SetHasBoundingBoxType` methods?

File: src/Camera_TEST.cc [hunk 12]
```diff
@@ -62,6 +62,18 @@ TEST(DOMCamera, Construction)
   cam.SetFarClip(200.2);
   EXPECT_DOUBLE_EQ(200.2, cam.FarClip());
 
+  cam.SetSegmentationType("semantic");
```
Comment (adlarkin) [inline src/Camera_TEST.cc:65]:
```suggestion
  EXPECT_EQUAL("semantic", cam.SegmentationType());
  cam.SetSegmentationType("panoptic");
```
The segmentation type should be `semantic` by default according to `sdf/1.9/camera.sdf`, so, can you check that the default value is correct and then test setting it to something that's not the default value?

File: src/Camera_TEST.cc [hunk 13]
```diff
@@ -62,6 +62,18 @@ TEST(DOMCamera, Construction)
...
+  EXPECT_EQ("semantic", cam.SegmentationType());
+  EXPECT_TRUE(cam.HasSegmentationType());
+
+  cam.SetBoundingBoxType("3d");
```
Comment (adlarkin) [inline src/Camera_TEST.cc:69]:
```suggestion
  EXPECT_EQ("2d", cam.BoundingBoxType())
  cam.SetBoundingBoxType("3d");
```
The bbox type is `2d` by default according to `sdf/1.9/camera.sdf`, so can you check that the default value is correct first before setting it to a different type?

File: test/integration/link_dom.cc [hunk 14]
```diff
@@ -358,6 +358,31 @@ TEST(DOMLink, Sensors)
...
+    link->SensorByName("segmentation_sensor");
+  ASSERT_NE(nullptr, segmentationSensor);
+  EXPECT_EQ("segmentation_sensor", segmentationSensor->Name());
+  EXPECT_EQ(sdf::SensorType::SEGMENTATION_CAMERA, segmentationSensor->Type());
+  EXPECT_EQ(ignition::math::Pose3d(37, 38, 39, 0, 0, 0),
+            segmentationSensor->RawPose());
+  const sdf::Camera *segmentationCameraSensor =
+    segmentationSensor->CameraSensor();
+  ASSERT_NE(nullptr, segmentationCameraSensor);
+  EXPECT_EQ("my_segmentation_camera", segmentationCameraSensor->Name());
```
Comment (adlarkin) [inline test/integration/link_dom.cc:372]:
```suggestion
  EXPECT_EQ("my_segmentation_camera", segmentationCameraSensor->Name());
  EXPECT_TRUE(segmentationCameraSensor->HasSegmentationType());
  EXPECT_EQ("semantic", segmentationCameraSensor->SegmentationType());
```
Now that the segmentation type is a part of the `sdf::Camera` API, can you add a check for the camera's segmentation type?

File: test/integration/link_dom.cc [hunk 15]
```diff
@@ -358,6 +358,31 @@ TEST(DOMLink, Sensors)
...
+  const sdf::Sensor *boundingboxSensor =
+    link->SensorByName("boundingbox_sensor");
+  ASSERT_NE(nullptr, boundingboxSensor);
+  EXPECT_EQ("boundingbox_sensor", boundingboxSensor->Name());
+  EXPECT_EQ(sdf::SensorType::BOUNDINGBOX_CAMERA, boundingboxSensor->Type());
+  EXPECT_EQ(ignition::math::Pose3d(37, 38, 39, 0, 0, 0),
+            boundingboxSensor->RawPose());
+  const sdf::Camera *boundingboxCamSensor = boundingboxSensor->CameraSensor();
+  ASSERT_NE(nullptr, boundingboxCamSensor);
+  EXPECT_EQ("my_boundingbox_camera", boundingboxCamSensor->Name());
```
Comment (adlarkin) [inline test/integration/link_dom.cc:384]:
```suggestion
  EXPECT_EQ("my_boundingbox_camera", boundingboxCamSensor->Name());
  EXPECT_TRUE(boundingboxCamSensor->HasBoundingBoxType());
  EXPECT_EQ("2d", boundingboxCamSensor->BoundingBoxType());
```
Now that the bbox type is a part of the sdf::Camera API, can you add a check for the camera's bbox type?

File: src/Camera.cc [hunk 16]
```diff
@@ -277,6 +289,16 @@ Errors Camera::Load(ElementPtr _sdf)
         "Camera sensor is missing a <clip> element."});
   }
 
+  if (_sdf->HasElement("segmentation_type"))
+  {
+    this->SetSegmentationType(_sdf->Get<std::string>("segmentation_type"));
```
Comment (adlarkin) [inline src/Camera.cc:294]:
```suggestion
    this->dataPtr->hasSegmentationType = true;
    this->SetSegmentationType(_sdf->Get<std::string>("segmentation_type"));
```

File: src/Camera.cc [hunk 17]
```diff
@@ -277,6 +289,16 @@ Errors Camera::Load(ElementPtr _sdf)
...
+  }
+
+  if (_sdf->HasElement("box_type"))
+  {
+    this->SetBoundingBoxType(_sdf->Get<std::string>("box_type"));
```
Comment (adlarkin) [inline src/Camera.cc:299]:
```suggestion
    this->dataPtr->hasBoundingBoxType = true;
    this->SetBoundingBoxType(_sdf->Get<std::string>("box_type"));
```

File: src/Camera.cc [hunk 18]
```diff
@@ -78,6 +78,12 @@ class sdf::Camera::Implementation
   /// \brief Far clip distance for depth camera.
   public: double depthFarClip{10.0};
 
+  /// \brief Segmentation type for segmentation camera.
+  public: std::string segmentationType;
+
+  /// \brief Boundingbox type for boundingbox camera.
+  public: std::string boundingBoxType;
```
Comment (adlarkin) [inline src/Camera.cc:85]: these variables should be initialized to their default values that were specified in `sdf/1.9/camera.sdf`

**Review by adlarkin — APPROVED** [review]
Nice work!

**Review by azeey — APPROVED** [review]
Looks good! Just a couple nits.

**Review by azeey — APPROVED** [review]

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/gsoc-2021-machine-learning-extension-to-ignition-gazebo/1070/1
